server/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```diff
@@ -417,89 +466,91 @@ private void loadAllLookupsAndInitStateRef()
+        LOG.info("Round of attempts #%d, [%d] lookups", i + 1, remainingLookups.size());
+        Map<String, LookupExtractorFactoryContainer> successfulLookups =
+            startLookups(remainingLookups, completionService);
+        builder.putAll(successfulLookups);
+        List<LookupBean> failedLookups = remainingLookups
+            .stream()
+            .filter(l -> !successfulLookups.containsKey(l.getName()))
+            .collect(Collectors.toList());
+        if (failedLookups.isEmpty()) {
+          break;
+        } else {
+          // next round
+          remainingLookups = failedLookups;
         }
       }
       stateRef.set(new LookupUpdateState(builder.build(), ImmutableList.of(), ImmutableList.of()));
     }
-    catch (Exception e) {
+    catch (InterruptedException | RuntimeException e) {
       LOG.error(e, "Failed to finish lookup load process.");
-      for (Future future : futures) {
-        future.cancel(true);
-      }
     }
     finally {
       executorService.shutdownNow();
     }
   }
 
+  /**
+   * @return a map with successful lookups
+   */
+  private Map<String, LookupExtractorFactoryContainer> startLookups(
+      List<LookupBean> lookupBeans,
+      CompletionService<Map.Entry<String, LookupExtractorFactoryContainer>> completionService
+  ) throws InterruptedException
+  {
+    for (LookupBean lookupBean : lookupBeans) {
+      completionService.submit(() -> startLookup(lookupBean));
+    }
+    Map<String, LookupExtractorFactoryContainer> successfulLookups = new HashMap<>();
+    for (int i = 0; i < lookupBeans.size(); i++) {
+      Future<Map.Entry<String, LookupExtractorFactoryContainer>> completedFuture = completionService.take();
+      try {
+        Map.Entry<String, LookupExtractorFactoryContainer> lookupResult = completedFuture.get();
+        if (lookupResult != null) {
+          successfulLookups.put(lookupResult.getKey(), lookupResult.getValue());
+        }
+      }
+      catch (ExecutionException e) {
+        LOG.error(e.getCause(), "Exception while starting a lookup");
+        // not adding to successfulLookups
+      }
+    }
+    return successfulLookups;
+  }
+
+  @Nullable
+  private Map.Entry<String, LookupExtractorFactoryContainer> startLookup(LookupBean lookupBean)
+  {
+    LookupExtractorFactoryContainer container = lookupBean.getContainer();
+    LOG.info("Starting lookup [%s]:[%s]", lookupBean.getName(), container);
+    try {
+      if (container.getLookupExtractorFactory().start()) {
+        LOG.info("Started lookup [%s]:[%s]", lookupBean.getName(), container);
+        return new AbstractMap.SimpleImmutableEntry<>(lookupBean.getName(), container);
+      } else {
+        LOG.error("Failed to start lookup [%s]:[%s]", lookupBean.getName(), container);
+        return null;
+      }
+    }
+    catch (RuntimeException e) {
+      throw new RE(e, "Failed to start lookup [%s]:[%s]", lookupBean.getName(), container);
```

Comment:
I changed from that version exactly to the current on @drcrallen's suggestion: https://github.com/druid-io/druid/pull/5029#discussion_r148146459

Reply:
i think @drcrallen thought that the exception would actually get propagated in some way and its not only for logging purpose... with exception it is doing the same thing of logging but a bit indirectly. for me, just logging and returning null would be simpler.
anyways, this is a nit and not really changes things in a big way. so, feel free to do whatever you two think to be good.

Reply:
My reasoning behind throwing instead of logging here is that this is wrapped in a retry loop. As such, it should be the responsibility of the retry loop to determine the severity of such an error. Options include: printing as an ERROR level log, printing as a WARN level log and retry, suppress the first occurrence and only print log messages on multiple failures, etc...

Basically since this is a retry-able thing, it is beyond the scope of just this method to determine how "noisy" the failure should be.